### PR TITLE
Added ability to update snaps when installing them

### PR DESCRIPTION
**Pull request by ritave**
This PR now has additional user approval that is not implemented in metamask-extension yet

### CODE REVIEW

File: packages/controllers/src/snaps/SnapController.ts [hunk 1]
```diff
@@ -1027,27 +1027,51 @@ export class SnapController extends BaseController<
...
+
+    const existingSnap = this.getTruncated(snapId);
+    // For devX we always re-install local snaps.
+    if (existingSnap && getSnapPrefix(snapId) !== SnapIdPrefixes.local) {
+      if (satisfiesSemver(existingSnap.version, versionRange)) {
+        return existingSnap;
+      }
+
+      try {
+        const updateResult = await this.updateSnap(
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1058]: I think we should have an accompanying UI PR to add an update prompt for this?

File: packages/controllers/src/snaps/SnapController.ts [hunk 2]
```diff
@@ -1027,27 +1027,51 @@ export class SnapController extends BaseController<
...
+   * @param versionRange - The semver range of the snap to install.
    * @returns The resulting snap object, or an error if something went wrong.
    */
   private async processRequestedSnap(
     origin: string,
-    snapId: SnapId,
-    version: string,
+    _snapId: SnapId,
+    versionRange: string,
   ): Promise<ProcessSnapResult> {
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1038]: I wonder if there is a way to simplify this function. with these additions there are now two paths that both call `authorize` etc. Do you think it's worth it to combine it somehow?

Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:1038]: As in `updateSnap` calls it and the remaining parts of this function calls it too

Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:1038]: Also does `startSnap` etc

File: packages/controllers/src/snaps/SnapController.ts (hunk 1 above)
Comment (rekmarks) [reply packages/controllers/src/snaps/SnapController.ts:1058]: Yes, we should ask the user before attempting to update a snap.

Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:1075]: Added a request for approval in a new commit

File: packages/controllers/src/snaps/SnapController.ts (hunk 2 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:1055]: I looked at it, and couldn't find any nice way to abstract it out cleanly.

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**